src/identify/index.js

```diff
@@ -1,7 +1,289 @@
+   * with the provided `protocols`
+   * @static
+   * @param {PeerInfo} peerInfo
+   * @param {Array<string>} protocols
+   */
+  static updatePeerProtocols (peerInfo, protocols) {
+    if (protocols && protocols.length > 0) {
+      peerInfo.protocols.clear()
+      protocols.forEach(proto => peerInfo.protocols.add(proto))
+    }
+  }
+
+  /**
+   * Takes the `addr` and converts it to a Multiaddr if possible
+   * @param {Buffer|String} addr
+   * @returns {Multiaddr|null}
+   */
+  static getCleanMultiaddr (addr) {
+    if (addr && addr.length > 0) {
+      try {
+        return multiaddr(addr)
+      } catch (_) {
+        return null
+      }
+    }
+    return null
+  }
+
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {Dialer} options.registrar
+   * @param {Map<string, handler>} options.protocols A reference to the protocols we support
+   * @param {PeerInfo} options.peerInfo The peer running the identify service
+   */
+  constructor (options) {
+    /**
+     * @property {Registrar}
+     */
+    this.registrar = options.registrar
+    /**
+     * @property {PeerInfo}
+     */
+    this.peerInfo = options.peerInfo
+
+    this._protocols = options.protocols
+
+    this.handleMessage = this.handleMessage.bind(this)
+  }
+
+  /**
+   * Send an Identify Push update to the list of peers
+   * @param {Array<PeerInfo>} peers
+   * @returns {Promise<void>}
+   */
+  push (peers) {
+    const pushes = peers.map(async peerInfo => {
+      // Don't push to peers who dont support it
+      if (!peerInfo.protocols.has(MULTICODEC_IDENTIFY_PUSH)) {
+        return
+      }
+
+      const connection = this.registrar.getConnection(peerInfo)
+      try {
+        const { stream } = await connection.newStream(MULTICODEC_IDENTIFY_PUSH)
+
+        await pipe(
+          [{
+            listenAddrs: this.peerInfo.multiaddrs.toArray().map((ma) => ma.buffer),
+            protocols: Array.from(this._protocols.keys())
+          }],
+          pb.encode(Message),
+          stream
+        )
+      } catch (err) {
+        // Just log errors
+        log.error('could not push identify update to peer', err)
+      }
+    })
+
+    return Promise.all(pushes)
+  }
+
+  /**
+   * Requests the `Identify` message from peer associated with the given `connection`.
+   * If the identified peer does not match the `PeerId` associated with the connection,
+   * an error will be passed to the `callback`.
+   *
+   * @async
+   * @param {Connection} connection
+   * @param {PeerID} expectedPeer The PeerId the identify response should match
+   * @returns {Promise<void>}
+   */
+  async identify (connection, expectedPeer) {
+    const { stream } = await connection.newStream(MULTICODEC_IDENTIFY)
+    const [data] = await pipe(
+      stream,
+      lp.decode(),
+      take(1),
+      toBuffer,
+      collect
+    )
+
+    if (!data) {
+      throw errCode('No data could be retrieved', codes.ERR_CONNECTION_ENDED)
```

Comment:
The stack trace should point here
```suggestion
      throw errCode(new Error('No data could be retrieved'), codes.ERR_CONNECTION_ENDED)
```